src/servant/tt_002/3678.c

```diff
@@ -657,9 +660,192 @@ void func_us_80175A78(Entity* self) {
     ServantUpdateAnim(self, NULL, D_us_80172B14);
 }
 
-INCLUDE_ASM("servant/tt_002/nonmatchings/3678", func_us_80175DBC);
+void func_us_80175DBC(Entity* self) {
+    s32 i;
+    s32 rnd;
+    s32 temp;
+
+    s_TargetLocOffset_calc = -0x18;
+    if (!PLAYER.facingLeft) {
+        s_TargetLocOffset_calc = -s_TargetLocOffset_calc;
+    }
+    s_TargetLocationX = PLAYER.posX.i.hi + s_TargetLocOffset_calc;
+    s_TargetLocationY = PLAYER.posY.i.hi - 0x18;
+    switch (self->step) {
+    case 0:
+        func_us_801739D0(self);
+        func_us_80173994(self, 0xE);
+        break;
+    case 1:
+        s_AngleToTarget =
+            CalculateAngleToEntity(self, s_TargetLocationX, s_TargetLocationY);
+        s_AllowedAngle = GetTargetPositionWithDistanceBuffer(
+            s_AngleToTarget, self->ext.faerie.targetAngle, 0x180);
+        self->ext.faerie.targetAngle = s_AllowedAngle;
+        self->velocityY = -(rsin(s_AllowedAngle) << 5);
+        self->velocityX = rcos(s_AllowedAngle) << 5;
+        func_us_80173BD0(self);
+        self->posX.val += self->velocityX;
+        self->posY.val += self->velocityY;
+        s_DistToTargetLocation =
+            CalculateDistance(self, s_TargetLocationX, s_TargetLocationY);
+        if (s_DistToTargetLocation < 2) {
+            self->step++;
+        }
+        break;
+    case 2:
+        self->facingLeft = PLAYER.facingLeft;
+        if (!g_Status.equipHandCount[D_us_80172494[self->params * 4 + 1]]) {
+            func_us_80173994(self, 0x10);
+            self->step = 5;
+            break;
+        }
+        if (SearchForEntityInRange(1, 0x27)) {
+            self->entityId = ENTITY_ID_SERVANT;
+            self->step = 0;
+            return;
+        }
+
+        func_us_80173994(self, 0x12);
+        self->step++;
+        break;
+    case 3:
+        if (PLAYER.facingLeft)
+            temp = 0;
+        else
+            temp = 1;
+        self->facingLeft = temp;
+
+        if (self->animFrameIdx == 0xB) {
+            for (rnd = rand() % 0x100, i = 0; true; i++) {
+                if (rnd <= D_us_80172BE4[i * 2]) {
+                    g_api.PlaySfx(D_us_80172BE4[(i * 2) + 1]);
+                    break;
+                }
+            }
+
+            g_Status.equipHandCount[D_us_80172494[self->params * 4 + 1]]--;
+
+            g_api.CreateEntFactoryFromEntity(
+                self, 0x37 + (D_us_80172494[self->params * 4 + 2] << 0x10), 0);
```

Comment:
I was going to do this, but my brain didn't process that 0x10 = 16.  For some reason I thought this was different that FACTORY and was only shifting 10 bytes instead of 16.  I'll fix this.